Vision lineup

linglejack06 opened the pull request:
- add cross track distance error method
- add along track distance error method
- add single tag observation record for helping with observations

linglejack06 commented:
@jkleiber could you check over my math calculations in the methods in vision localizer?

also... for the cross track error, how will we determine left/right? how its calculated now would always be a positive number i think? unless target yaw goes negative when camera is to the right of target?

aidnem commented on vision/src/main/java/coppercore/vision/VisionLocalizer.java:
```diff
@@ -64,6 +66,72 @@ public VisionLocalizer(
+        }
+
+        // we can subtract the angle of camera if there is one
+        return tagObserved.distance3D() * Math.sin(tagObserved.tx().minus(new Rotation2d()).getRadians())
```
Why does this .minus(new Rotation2d())? Wouldn't that just be zero?

linglejack06 replied:
Yes @aidnem I added it as an example. We would need a rotation 2d angle there in cases where our camera is rotated inwards or outwards / upwards or downwards

So the new rotetion2d is a placeholder to explain that that might be needed

jkleiber commented on vision/src/main/java/coppercore/vision/VisionLocalizer.java:
```diff
@@ -64,6 +66,72 @@ public VisionLocalizer(
+    public double getCrossTrackDistanceErrorToTag(
+            int tagId, int desiredCameraIndex, double crossTrackOffsetMeters) {
+        // camera not in vision
+        if (desiredCameraIndex >= inputs.length) {
```
Rather than return -1 here or below, we should do one of the following things:

1. Return a huge number and then only accept values within a certain error range (otherwise we do nothing or use the last value or something)

2. Make a new data type with value and validity as fields (possible coppercore future feature)


You can pick, idc which we choose

linglejack06 replied:
Ok I can probably just make another record

jkleiber commented on vision/src/main/java/coppercore/vision/VisionLocalizer.java:
```diff
@@ -64,6 +66,72 @@ public VisionLocalizer(
+        }
+
+        // we can subtract the angle of camera if there is one
+        return tagObserved.distance3D() * Math.sin(tagObserved.tx().minus(new Rotation2d()).getRadians())
```
What is tx()?

Could we make this function name more obvious? As the math is written, this has to be an angle to the tag from the camera. We can probably come up with something better (headingToTag ?)

jkleiber replied:
Actually okay so ty is vertical angle to tag and tx is yaw angle to tag based on my vague/weak understanding of this. If that's the case wouldn't taking cos of either tx / ty not be quite accurate?

To demonstrate this, consider a scenario where the tag is dead center on the camera in X (tx=0) but ty=45 degrees. Then:
- using tx only: dist*cos(0) is incorrect because we leave out the vertical angle
- using ty only: dist*cos(45 deg) is correct because we factor in the angle

But if we are dead center vertically but the tag is 45 degrees to one side:
- using tx only: dist*cos(45 deg) is right because we have no vertical error but do have a yaw error 
- using ty only: dist*cos(0) is wrong as this is just the hypotenuse

Another way to think of it: The distance to the tag is represented by a line in 3d space, so both angles matter.

This almost certainly impacts the crosstrack error equation as well. I am on my phone so deriving this is too tough rn but I'll think about it some more. Give it a shot though

jkleiber commented on vision/src/main/java/coppercore/vision/VisionLocalizer.java:
```diff
@@ -64,6 +66,72 @@ public VisionLocalizer(
+        }
+
+        // we can subtract the angle of camera if there is one
+        return tagObserved.distance3D() * Math.cos(tagObserved.ty().minus(new Rotation2d()).getRadians())
```
What is ty?

Since we're doing cos() here I think we can use the same angle as used above.

My concern is that tx and ty aren't actually angles. In which case we should figure out what they are

jkleiber reviewed (commented):
Math feedback - I'll be at shop starting from around 12 and we can try working this out

jkleiber commented:
I think these are the correct equations:

xtk error = 3d_dist * cos(vertical angle) * sin(yaw angle)

forward distance= 3d_dist * cos(vertical angle) * cos(yaw angle)

If yaw angle is signed then we'll have a signed xtk error.

linglejack06 commented:
@jkleiber those equations probably make more sense, for forward distance though, if we take cos of vertical and cos of horizontal, aren't we then getting to forward distances away? (For times when one isn't 0?) I might just need to work it out on paper to make exact sense of your equations though as I'm sure they work

linglejack06 commented:
Is this similar to your equation: [3d position line](https://stackoverflow.com/questions/32551940/find-a-new-3d-position-from-a-known-point-of-origin-based-on-distance-and-angle)

jkleiber commented:
@linglejack06 I'm not sure what you mean by forward distances away. But my goal with the forward distance calculation is to get the distance between the robot and the reef side along the vector perpendicular to the reef (along-track distance)

In my forward distance equation, it can only be 0 if 3d_dist is 0 (since we won't be able to observe the tag at 90 degrees). For the xtk error, it'll be 0 if 3d_dist is 0 or if we are centered. So we shouldn't modify the 3d_dist measurement to correct for camera position before computing xtk error (we should apply any offset afterwards)

It does look like the equations I posted are similar to radius-azimuth-elevation, I just defined the angles slightly different

linglejack06 commented:
That’s ridiculous lolSent from my iPhoneOn Jan 25, 2025, at 4:29 PM, Justin Kleiber ***@***.***> wrote:﻿
@linglejack06 shortly after you left we got the JSON constants loading, and I just pushed to the test-integration branch
@avidraccoon identified the problem as being the need for using Double rather than double.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

linglejack06 commented:
@jkleiber i updated calculations and switched to just one method that returns a record.

before we test and merge this, do you think its worthwhile to check if the tag observation is stale, ie if timestamp is too far behind robot timestamp, if so, would this just be passed into the method i assume (latest timestamp)

also... right now single tag observations are only gathered if there is no multitag result. I could also gether them by using the best target in the multitag result. This might be helpful if we are far to where another reef side tag accidentally gets into the frame.

in which case, the robot would recognize a multitag result, but we would still want the ability to use the best target (which should be the tag we are viewing straight on) to use our landing zone alignment?

linglejack06 commented:
let me know if my explanations regarding multitag result doesn't make sense, i dont think i worded it that well

jkleiber commented:
@linglejack06 I think if we're far enough away to see multiple tags we should be using global pose in any lineup logic

```
if (isMultitag) {
    // Compute offset distances from global pose
} else {
    // Compute offset via tag relative math
}
```

I think that logic would live in the drivetrain code though

linglejack06 commented:
> @linglejack06 I think if we're far enough away to see multiple tags we should be using global pose in any lineup logic
> 
> ```
> if (isMultitag) {
>     // Compute offset distances from global pose
> } else {
>     // Compute offset via tag relative math
> }
> ```
> 
> I think that logic would live in the drivetrain code though

ok, thats a good option, the multitag would be realyl easy too as id just use the pose gathered in drive with vision

linglejack06 commented:
@jkleiber this should be ready for review

jkleiber reviewed (approved):
Approved
